#!/usr/bin/env python3

"""
Script to generate filtered gene families from MMseqs2 clustering output.

Only strict 1-to-1 orthologous clusters are retained, meaning that each cluster must contain exactly one gene from each genome. Clusters that do not meet this criterion are skipped and logged.

Input:
- clusters_cluster.tsv generated by MMseqs2 (../results/clusters/mmseqs2/)

Output:
- orthologs1to1.tsv containing filtered 1-to-1 orthologous gene families

Paralogous clusters are filtered out by rejecting any cluster that contains more than one gene originating from the same genome.

6th part of pipeline
"""


# ~~~~~ Imports ~~~~~
from collections import defaultdict
import argparse
from pathlib import Path
from tqdm import tqdm

# ~~~~~ Paths ~~~~~
parser = argparse.ArgumentParser(
    description="Script to generate filtered gene families from MMseqs2 clustering output."
)

parser.add_argument(
    "--input",
    type=Path,
    required=True,
    help="Path to folder .tsv file *_cluster.tsv (MMseqs2 output)",
)

parser.add_argument("--output", required=True, type=Path)

args = parser.parse_args()
INPUT = Path(args.input)
OUTPUT = Path(args.output)
OUTPUT.parent.mkdir(parents=True, exist_ok=True)
LOG = Path(OUTPUT.parent / "filter_cluster.log")
LOG.write_text("")


# ~~~~~~ Functions ~~~~~
def species_from_id(gene_id: str) -> str:
    """Extract species name from gene ID."""
    return gene_id.split("|")[0]


def load_clusters(input_file: Path) -> dict:
    """Group genes by their representative sequence IDs into dictionary."""
    clusters = defaultdict(list)

    with input_file.open() as infile:
        for line in infile:
            rep_id, member_id = line.rstrip("\n").split("\t")
            clusters[rep_id].append(member_id)

    return clusters


if __name__ == "__main__":
    clusters = load_clusters(
        INPUT
    )  # read clusters from MMSeqs2 output: <member_sequence_id>    <representative_sequence_id>
    # count how many genomes are present
    all_species = set()
    for members in clusters.values():
        for gene_id in members:
            all_species.add(species_from_id(gene_id))

    N_GENOMES = len(all_species)

    with OUTPUT.open("w") as outfile:
        outfile.write("cluster_id\tspecie\tgene_id\n")

        for rep_id, members in tqdm(
            clusters.items(), desc="Filtering clusters", unit="clusters"
        ):
            species = [
                species_from_id(gene_id) for gene_id in members
            ]  # species present in the current cluster

            # Strict 1-1 condition - filter only those clusters which have exactly one gene from each genome
            if len(members) == N_GENOMES and len(set(species)) == N_GENOMES:
                for gene in members:
                    outfile.write(f"{rep_id}\t{species_from_id(gene)}\t{gene}\n")
            else:
                with LOG.open("a") as logfile:
                    logfile.write(
                        f"Cluster {rep_id} skipped: {len(members)} members from {len(set(species))} species (needed: {N_GENOMES})\n"
                    )
